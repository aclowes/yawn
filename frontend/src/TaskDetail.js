import React from 'react';
import {PanelGroup, Panel, Alert, Pagination, Button} from 'react-bootstrap';
import {Link} from 'react-router';

import API from "./API";

export default class TaskDetail extends React.Component {
  constructor(props) {
    super(props);
    this.state = {task: null, error: null};
  }

  componentDidMount() {
    API.get(`/api/tasks/${this.props.params.id}/`, (payload, error) => {
      const execution = payload ? payload.executions.length : null;
      this.setState({task: payload, error, execution});
    });
  }

  enqueue = () => {
    API.patch(`/api/tasks/${this.props.params.id}/`, {enqueue: true}, (payload, error) => {
      this.setState({task: payload, error});
    });
  };

  terminate = () => {
    const execution = this.state.task.executions[this.state.execution - 1];
    API.patch(`/api/tasks/${this.props.params.id}/`, {terminate: execution.id}, (payload, error) => {
      this.setState({task: payload, error});
    });
  };

  selectExecution = (eventKey) => {
    this.setState({
      execution: eventKey
    });
  };

  renderMessages() {
    return this.state.task.messages.map((message) => (
      <Link to={`/queue/${message.queue_id}`} key={message.id}>{message.queue}</Link>
    ))
  }

  renderWorkflowLink() {
    const task = this.state.task;
    if (task.workflow) return (
      <Link to={`/workflows/${task.workflow.id}`}>
        {task.workflow.name} - v{task.workflow.version}
      </Link>
    )
  }

  renderExecution() {
    if (this.state.execution === 0) {
      return <div>No executions</div>
    }
    const execution = this.state.task.executions[this.state.execution - 1];
    return (
      <dl className="dl-horizontal">
        <dt>Execution</dt>
        <dd>
          <Pagination
            ellipsis
            bsSize="small"
            items={this.state.task.executions.length}
            maxButtons={10}
            activePage={this.state.execution}
            onSelect={this.selectExecution}/>
        </dd>
        <dt>Status</dt>
        <dd>{execution.status}</dd>
        <dt>Worker</dt>
        <dd>
          <Link to={`/workers/${execution.worker.id}`}>
            {execution.worker.name}
          </Link>
        </dd>
        <dt>Start Time</dt>
        <dd>{execution.start_timestamp}</dd>
        <dt>Stop Time</dt>
        <dd>{execution.stop_timestamp}</dd>
        <dt>Exit Code</dt>
        <dd>{execution.exit_code}</dd>
        <dt>Standard Output</dt>
        <dd>
          <pre>{execution.stdout}</pre>
        </dd>
        <dt>Standard Error</dt>
        <dd>
          <pre>{execution.stderr}</pre>
        </dd>
      </dl>
    )
  }

  render() {
    if (this.state.task === null) {
      return (
        <div>
          {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
          Loading...
        </div>
      )
    } else {
      const task = this.state.task;
      return (
        <div>
          <PanelGroup>
            <Panel header="Details">
              <dl className="dl-horizontal">
                <dt>Workflow</dt>
                <dd>
                  {this.renderWorkflowLink()}
                </dd>
                <dt>Task Name</dt>
                <dd>{task.name}</dd>
                <dt>Command</dt>
                <dd>
                  <pre>{task.command}</pre>
                </dd>
                <dt>Max Retries</dt>
                <dd>{task.max_retries}</dd>
                <dt>Timeout</dt>
                <dd>{task.timeout} {task.timeout ? 'seconds' : '(none)'}</dd>
                <dt>Status</dt>
                <dd>{task.status}</dd>
                <dt>Queued Executions</dt>
                <dd>{this.renderMessages()}</dd>
              </dl>
              <Button onClick={this.enqueue}>Run</Button>
            </Panel>
            <Panel header="Executions">
              {this.renderExecution()}
              <Button onClick={this.terminate}>Terminate</Button>
            </Panel>
          </PanelGroup>
        </div>
      )
    }
  }
}
