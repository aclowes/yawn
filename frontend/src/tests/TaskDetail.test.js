import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import TaskDetail from '../TaskDetail';
import {mockAPI} from './mocks'

it('TaskDetail loading', () => {
  mockAPI([null]);
  const component = ReactTestRenderer.create(<TaskDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('TaskDetail failure', () => {
  mockAPI();
  const component = ReactTestRenderer.create(<TaskDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('TaskDetail success', () => {
  mockAPI([
    {
      "id": 1,
      "name": "start",
      "workflow": {"id": 1, "name": "Simple Workflow Example", "version": 1},
      "executions": [{
        "id": 1,
        "worker": {
          "id": 1,
          "name": "graybeard.local 35405",
          "status": "exited",
          "start_timestamp": "2017-01-30T03:26:38.197807Z",
          "last_heartbeat": "2017-01-30T03:44:38.063205Z"
        },
        "status": "succeeded",
        "start_timestamp": "2017-01-30T03:26:38.211839Z",
        "stop_timestamp": null,
        "exit_code": null,
        "stdout": "Starting...\n",
        "stderr": ""
      }],
      "messages": [],
      "max_retries": 0,
      "timeout": null,
      "command": "echo Starting...",
      "status": "succeeded"
    }
  ]);
  const component = ReactTestRenderer.create(<TaskDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
