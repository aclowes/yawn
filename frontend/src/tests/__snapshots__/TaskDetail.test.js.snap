// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDetail failure 1`] = `
<div>
  <div
    className="alert alert-danger"
    role="alert"
  >
    some error message
  </div>
  Loading...
</div>
`;

exports[`TaskDetail loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`TaskDetail success 1`] = `
<div>
  <div
    className="panel-group"
  >
    <div
      className="panel panel-default"
      id={undefined}
    >
      <div
        className="panel-heading"
      >
        Details
      </div>
      <div
        className="panel-body"
      >
        <dl
          className="dl-horizontal"
        >
          <dt>
            Workflow
          </dt>
          <dd>
            <a
              onClick={[Function]}
              style={Object {}}
            >
              Simple Workflow Example
               - v
              1
            </a>
          </dd>
          <dt>
            Task Name
          </dt>
          <dd>
            start
          </dd>
          <dt>
            Command
          </dt>
          <dd>
            "echo Starting..."
          </dd>
          <dt>
            Max Retries
          </dt>
          <dd>
            0
          </dd>
          <dt>
            Timeout
          </dt>
          <dd>
             
            (none)
          </dd>
          <dt>
            Status
          </dt>
          <dd>
            succeeded
          </dd>
          <dt>
            Queued Executions
          </dt>
          <dd />
        </dl>
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Run
        </button>
      </div>
    </div>
    <div
      className="panel panel-default"
      id={undefined}
    >
      <div
        className="panel-heading"
      >
        Executions
      </div>
      <div
        className="panel-body"
      >
        <dl
          className="dl-horizontal"
        >
          <dt>
            Execution
          </dt>
          <dd>
            <ul
              className="pagination pagination-sm"
            >
              <li
                className="active"
                style={undefined}
              >
                <a
                  href="#"
                  onClick={[Function]}
                  role="button"
                >
                  1
                </a>
              </li>
            </ul>
          </dd>
          <dt>
            Status
          </dt>
          <dd>
            succeeded
          </dd>
          <dt>
            Worker
          </dt>
          <dd>
            <a
              onClick={[Function]}
              style={Object {}}
            >
              graybeard.local 35405
            </a>
          </dd>
          <dt>
            Start Time
          </dt>
          <dd>
            2017-01-30T03:26:38.211839Z
          </dd>
          <dt>
            Stop Time
          </dt>
          <dd />
          <dt>
            Exit Code
          </dt>
          <dd />
          <dt>
            Standard Output
          </dt>
          <dd>
            <pre>
              Starting...
              
            </pre>
          </dd>
          <dt>
            Standard Error
          </dt>
          <dd>
            <pre>
              
            </pre>
          </dd>
        </dl>
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Terminate
        </button>
      </div>
    </div>
  </div>
</div>
`;
